compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -709,7 +731,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         });
 
         let ret_ty = ret_coercion.borrow().expected_ty();
-        let return_expr_ty = self.check_expr_with_hint(return_expr, ret_ty.clone());
+        // This is a coercion site, so we allow coercion of type ascriptions in return_expr
+        let expected = ExpectHasType(ret_ty.clone(), TypeAscriptionCtxt::Coercion);
+        let return_expr_ty = self.check_expr_with_expectation(return_expr, expected);
```

Comment:
Hmm, it seems like we were handling the coercion in a very particular way here. That is, there is a reason we didn't involve `check_expr_coercable_to`... I think that the reason is because we wanted to use a distinct `ObligationCauseCode` (`ObligationCauseCode::ReturnValue`) for better diagnostics. This makes me wonder if the "coercion site" type that you added would be better extended to carry a cause code, so we could rewrite this code to just invoke `check_expr_coercable_to...`